[ADD] estate: create estate module

rmbh-odoo opened the pull request:
After this PR:
- Added init and manifest file
- Added name, depends, and license in the manifest file

rmbh-odoo committed:
[ADD] estate: add estate module

After this commit:
- Added init and manifest file
- Added name and depends in the manifest file

rmbh-odoo committed:
[IMP] estate: update estate module

After this commit:
- Add Licence in manifest

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/280.png)](https://mergebot.odoo.com/odoo/tutorials/pull/280)

shag-odoo commented:
@rmbh-odoo PR title should be according to git guidelines
Reference : https://www.odoo.com/documentation/18.0/contributing/development/git_guidelines.html

shag-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,5 @@
+    'name': "estate",
+    'depends':['base'],
+    'license': 'LGPL-3',
+}
```
@rmbh-odoo new line missing at the end of file
refer codebase for file structure

rmbh-odoo committed:
[ADD] estate: add estate property model

After this commit:
- Created estate_property table with different attributes

shag-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,5 @@
+{
+    'name': "estate",
```
```suggestion
    'name': 'estate',
```

shag-odoo commented on estate/__manifest__.py:
```diff
@@ -1,5 +1,5 @@
     'license': 'LGPL-3',
-}
\ No newline at end of file
+}
```
```suggestion
}

```

shag-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,27 @@
+            ('west','West'),
+        ]
+    )
+
```
```suggestion
```

shag-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,21 @@
+    garden_area=fields.Integer()
+    garden_orientation=fields.Selection(
+        selection=[('north','North'),('south','South'),('east','East'),('west',)]
+    )
```
pending

shag-odoo commented on estate/__init__.py:
```diff
@@ -0,0 +1,2 @@
+from . import models
+
```
```suggestion
```

rmbh-odoo committed:
[ADD] estate: added security layer for access right

After this commit:
- Added ir.model.access.csv to give access rights to base.group_user

rmbh-odoo committed:
[ADD] estate: created user interface with different fields

After this commit:
- Added nested menu items to create a list and form view
- Added window action functionality with list and form view

rmbh-odoo committed:
[ADD] estate: added list, form and search view with filter and group

After this commit:
- The user can easily see data in the list and form view.
- Search data based on different record fields.
- Filter data based on availability.
- Group data based on postcode.

rmbh-odoo committed:
[ADD] estate: added relations beween model (Many2one, One2many, Many2many)

After this commit:
- The user can select a property type and add using the Many2one relation model.
- User can attach different tags with the property.
- Seller can see different offers in the property form itself.

shag-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,63 @@
+    state=fields.Selection(
+        required=True,
+        default='new',
+        copy=False,
```
```suggestion
        copy=False,
```
What is the purpose of this attribute?

rmbh-odoo replied:
In Action Gear, duplicating a form doesn't copy this attribute value.

shag-odoo replied:
it also resist copying data of the field
you can resolve it

shag-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,63 @@
+        ],
+    )
+    property_type_id=fields.Many2one(
+        comodel_name='estate.property.type',
```
```suggestion
        comodel_name='estate.property.type',
```
IS it necessary to add comodel name?

rmbh-odoo replied:
comodel_name is mandatory (except for related or extended fields)
Example:

```
property_type_id = fields.Many2one(
    'estate.property.type',  # comodel_name (must be first)
    'Property Type',        # string (label)
)
```

If multiple positional arguments are used, they must follow the function's parameter order. You can't write 'Property Type' before 'estate.property.type' without explicitly specifying parameter labels, as per Python’s function argument rules.

shag-odoo replied:
It is not compulsory to explicity mention `comodel_name` you can directly write the model name but it should be written first
Ex- `property_type_id = fields.Many2one('estate.property.type', string="Type")` -> It will work
 `property_type_id = fields.Many2one(string="Type", 'estate.property.type',)` -> It will not work
 
 you can resolve it

shag-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,63 @@
+        comodel_name='estate.property.type',
+        string='Property Type',
+    )
+    salesman=fields.Many2one(
```
```suggestion
    user_id = fields.Many2one(
```
It is more relevant.

shag-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,63 @@
+        string='Salesman',
+        default=lambda self: self.env.user,
+    )
+    buyer=fields.Many2one(
```
```suggestion
    partner_id = fields.Many2one(
```

shag-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,63 @@
+    )
+    offer_ids=fields.One2many(
+        comodel_name="estate.property.offer",
+        inverse_name='property_id',
```
```suggestion
        inverse_name='property_id',
```
what is the use case of inverse_name and what happens if it is not defined?

rmbh-odoo replied:
inverse_name is required in One2many fields to map the relationship to a corresponding Many2one field in the target model.

Without it, Odoo can't determine how records in the target model relate to each other, causing a no inverse field error. 

This ensures proper bidirectional access and data integrity.

shag-odoo replied:
resolve it

shag-odoo commented on estate/models/estate_property_tag.py:
```diff
@@ -0,0 +1,7 @@
+from odoo import fields,models
+
+class EsatePropertyTag(models.Model):
+    _name='estate.property.tag'
+    _description='Estate Property Tag'
+
+    name=fields.Char()
```
```suggestion
    _name = 'estate.property.tag'
    _description = 'Estate Property Tag'

    name = fields.Char()
```
check the definition of field declaration in the codebase
for every model you have created in your module this problem exists
Please fix this

shag-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,96 @@
+                <field name="property_type_id"/>
+                <filter string="Available" name="available_filter" domain="['|',('state','=','new'),('state','=','offer_received')]"/>
+                <group string="Group By">
+                    <filter string="Postcode" name="postcode_group" context="{'group_by':'postcode'}"/>
```
```suggestion
                    <filter string="Postcode" name="postcode_group" context="{'group_by':'postcode'}"/>
```
what is the purpose of context?

rmbh-odoo replied:
Context is used to pass additional parameters that influence views, actions, or methods dynamically.

Here, context="{'group_by':'postcode'}" ensures records are grouped by postcode when this filter is applied.

In the codebase: [models.py#L2891](https://github.com/odoo/odoo/blob/18.0/odoo/models.py#L2891),
group_by can be accessed on the backend to return records accordingly.

shag-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,96 @@
+                    <notebook>
+                        <page string="Description">
+                            <group>
+                                <field name="description" placeholder="write sort description about property" string="Description"/>
```
```suggestion
                                <field name="description" placeholder="Write short description about property" string="Description"/>
```

shag-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,96 @@
+            <list string="Property">
+                <field name="name"/>
+                <field name="property_type_id"/>
+                <field name="tag_ids" widget="many2many_tags"/>
```
```suggestion
                <field name="tag_ids" widget="many2many_tags"/>
```
what if widget is not defined
try to check the code of many2many_tags in code base and update the file path in PR comment

rmbh-odoo replied:
Normally, a many2many field is displayed as a list or a many-to-many table.
Using widget="many2many_tags", the field is displayed as colored tags instead.

file path: [addons/web/static/src/views/fields/many2many_tags/many2many_tags_field.js](https://github.com/odoo/odoo/blob/18.0/addons/web/static/src/views/fields/many2many_tags/many2many_tags_field.js)

in 292 line many2many_tag register with many2ManyTagsField function as a category

shag-odoo replied:
resolve it

rmbh-odoo committed:
[FIX] estate: fix spacing in attribute, variable name, import order

After this commit:
- Fix spacing in variables according to the  codebase
- Fix import library order according to coding guidelines
- Fix variable name

rmbh-odoo committed:
[ADD] estate: compute and onchange Enhancements

After this commit:
- best_offer: Auto-computes the highest offer price.
- date_deadline: Auto-computes based on validity if not set manually.
- garden: Auto-sets garden_area and garden_orientation by default.

rmbh-odoo committed:
[ADD] estate: add Actions for Sale, Cancellation, and Offer Management

After this commit:
- Users can mark a property as sold or canceled.
- A sold property can't be canceled, and vice versa.
- Added Accept and Refuse buttons for offers:
    - Accept: Updates offer status, sets selling_price, and assigns buyer.
    - Refuse: Marks the offer as refused.

rmbh-odoo committed:
[ADD] estate: add Kanban View for Properties

After this commit:
- Users can view and edit properties in the Kanban view.

rmbh-odoo committed:
[ADD] estate: add SQL & Python Constraints

After this commit:
- Prevents negative values for selling_price and expected_price.
- Enforces unique names for property_type and property_tag.
- Restricts offer acceptance if the price is below 90% of the expected price.

shag-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,116 @@
+        string='Property Type',
+    )
+    user_id = fields.Many2one(
+        comodel_name='res.users',
```
```suggestion
'res.users',
```
you can remove comodel_name label/attr from every relational field

shag-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,116 @@
+    @api.onchange('garden')
+    def _onchange_garden(self):
+        if self.garden == True:
+            self.garden_area = 10
```
```suggestion
            self.garden_area = 10
```
why static value?

rmbh-odoo replied:
In the tutorial, it is mentioned that a static value (self.garden_area = 10) is used because the garden area is set to a default value of 10 whenever the garden option is clicked.

shag-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,116 @@
+    def check_selling_price(self):
+        for record in self:
+            if not float_is_zero(record.selling_price, precision_digits=1) and float_compare(record.selling_price, 0.9 * record.expected_price, precision_digits=1) == -1:
+                raise ValidationError('The selling price must be at least 90% of the expected price! You must reduce expected price if you want to accept this offer')
```
```suggestion
                raise ValidationError(_('The selling price must be at least 90% of the expected price! You must reduce expected price if you want to accept this offer'))
```
Use translation here also

shag-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,57 @@
+        for record in self:
+            if record.validity and record.property_id.create_date:
+                record.date_deadline = record.property_id.create_date + timedelta(days=record.validity)
+            else:
+                record.date_deadline = False
```
can't we set dealdline to `today + validity `

rmbh-odoo committed:
[ADD] estate: ui Enhancements & Improved Functionality

After this commit:
- View properties directly from the property type model.
- Custom list order for better organization.
- Conditional display of buttons, fields, and tag colors.
- Inline editing of offers and tags in the list view.
- Filter for living_area to show results larger than a specified value.
- Stat button in property type form view to display all related offers.

rmbh-odoo committed:
[ADD] estate: offer validation and property visibility with inheritance

After this commit:
- Users can only place offers that exceed the current best offer.
- Salespersons can view properties assigned to them directly from their profile.

rmbh-odoo committed:
[ADD] estate: auto-generate invoice on property sale

After this commit:
- Automatically creates an invoice when a property is sold.
- Requires the Invoice app to be installed.

rmbh-odoo committed:
[IMP] estate: enhance Kanban View for Better UI

After this commit:
- Properties are grouped by their type for better organization.
- Property status is displayed using a progress bar for clear visibility.

rmbh-odoo committed:
[ADD] estate: restrict access to data

After this commit:
- Set CRUD permissions based on user groups.
- Limit agents to viewing properties.
- Secure sudo usage with access checks.
- Support multi-company with proper security.
- Restrict agents from accessing property settings.

rmbh-odoo committed:
[ADD] estate: add default data and demo records

After this commit:
- Created default property types and tags.
- Added demo property and offer data for user convenience.

rmbh-odoo committed:
[ADD] estate: add property and salesman reports

After this commit:
- Allow users to download property reports with details and offers.
- Enable salesmen to download reports of all properties with their offers.
- View property invoice details directly if the property is sold.

rmbh-odoo committed:
[ADD] estate: implement property listing and detail view on website

After this commit:
- Enabled users to view available properties through dedicated URLs
- Users can now view detailed information and offers for each property
- Integrated subcomponents to display property effectively on the website
- Added additional demo data to enhance user experience and testing

rmbh-odoo committed:
[ADD] awesome_owl: introduce sum, card, and dynamic todo list features in OWL

After this commit:
- Implemented a counter with an increment button to display the total sum,
- Enabled dynamic property passing for the card component.
- Developed a dynamic to-do list allowing users to create and edit tasks.

rmbh-odoo committed:
[ADD] awesome_owl: complete TODO functionality with subcomponent integration

After this commit:
- Users can create and delete TODO items.
- Added functionality to mark TODO items as done.
- Passed counter subcomponent into the card component.
- Add chatter in estate property.

rmbh-odoo committed:
[ADD] awesome_dashboard: add quick navigation and dynamic cards

After this commit:
- Added quick navigation buttons for Customers and Leads.
- Displayed cards with dynamic data fetched via RPC call to the controller.

rmbh-odoo commented:
Closing this as it’s part of my training task.
